crates/downloader/src/download_queue.rs

```diff
@@ -0,0 +1,283 @@
+use std::{
+    collections::BTreeMap,
+    future::Future,
+    time::Duration,
+};
+
+use alloy::{
+    eips::BlockNumberOrTag,
+    providers::{
+        Provider,
+        RootProvider,
+    },
+    rpc::types::Block,
+    transports::{
+        http::Http,
+        RpcError,
+        TransportErrorKind,
+    },
+};
+use futures_util::Stream;
+use tokio::{
+    sync::mpsc,
+    task::JoinSet,
+};
+use tokio_stream::wrappers::ReceiverStream;
+
+#[derive(Debug)]
+pub enum GetBlockError {
+    RateLimited,
+    Other(RpcError<TransportErrorKind>),
+}
+
+pub trait GetBlock {
+    fn get_block(
+        &self,
+        block_number: u64,
+    ) -> impl Future<Output = Result<Option<Block>, GetBlockError>> + Send;
+}
+
+impl GetBlock for RootProvider<Http<reqwest::Client>> {
+    async fn get_block(&self, block_number: u64) -> Result<Option<Block>, GetBlockError> {
+        self.get_block_by_number(BlockNumberOrTag::Number(block_number), true)
+            .await
+            .map_err(|err| match err {
+                RpcError::ErrorResp(resp) if resp.code == 429 => {
+                    GetBlockError::RateLimited
+                }
+                other => GetBlockError::Other(other),
+            })
+    }
+}
+
+enum Mode {
+    /// Batch-download blocks until we reach the end of the chain
+    Batch,
+    /// Poll for new blocks
+    Poll,
+}
+impl Mode {
+    /// How many blocks to download concurrently
+    fn concurrency(&self) -> usize {
+        match self {
+            Mode::Batch => 8,
+            Mode::Poll => 1,
+        }
+    }
+}
+
+pub struct DownloadQueue<P> {
+    provider: P,
+    next_da_height: u64,
+    pending: JoinSet<Result<Option<Block>, GetBlockError>>,
+    completed: BTreeMap<u64, Block>,
+}
+
+impl<P> DownloadQueue<P>
+where
+    P: GetBlock + Clone + Send + 'static,
+{
+    pub fn start_from(provider: P, da_height: u64) -> Self {
+        Self {
+            provider,
+            next_da_height: da_height,
+            pending: JoinSet::new(),
+            completed: BTreeMap::new(),
+        }
+    }
+
+    async fn stream_inner(
+        mut self,
+        tx: mpsc::Sender<Result<Block, anyhow::Error>>,
+    ) -> anyhow::Result<()> {
+        let mut next_to_emit = self.next_da_height;
+        let mut mode = Mode::Batch;
+        const DA_BLOCK_TIME_MILLIS: u64 = 12_000;
+
+        loop {
+            // Queue new downloads up to the concurrency limit
+            while self.pending.len() + self.completed.len() < mode.concurrency() {
+                let provider = self.provider.clone();
+                self.pending
+                    .spawn(async move { provider.get_block(self.next_da_height).await });
+                self.next_da_height += 1;
```

Comment:
Shouldn't we have a lint to prevent unsafe arithmetic? Can be done as a follow-up.